### PR TITLE
Refactor Parser, update dependencies

### PR DESCRIPTION
**Pull request by ganchclub**

### CODE REVIEW

File: src/ElementFinderFactory.php [hunk 1]
```diff
@@ -1,37 +1,68 @@
...
+
+    /**
+     * @param LinkConverterInterface|null $linkConverter
+     * @param EncodingConverterInterface|null $encodingConverter
+     * @param ExpressionTranslatorInterface|null $expressionTranslator
+     */
+    public function __construct(
+      LinkConverterInterface $linkConverter = null,
+      EncodingConverterInterface $encodingConverter = null,
+      ExpressionTranslatorInterface $expressionTranslator = null
```
Comment (funivan) [inline src/ElementFinderFactory.php]: Maybe expression translator we should set as the first argument ?

File: src/ElementFinderFactory.php [hunk 2]
```diff
@@ -1,37 +1,68 @@
...
+    ) {
+      if ($linkConverter === null) {
+        $this->linkConverter = new LinkConverter();
+      }
+      if ($encodingConverter === null) {
+        $this->encodingConverter = new HtmlEncodingConverter();
+      }
+      $this->expressionTranslator = new XpathExpression();
```
Comment (funivan) [inline src/ElementFinderFactory.php]: bug. Parameter not used `$expressionTranslator`

File: src/Helper/EncodingConverterInterface.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  interface EncodingConverterInterface {
+
+    /**
+     * Try to detect input encoding from contentType or from html <meta> tag
+     *
+     * @param string $html
+     * @param string $contentType
+     * @return string
+     */
+    public function convertToUtf(string $html, string $contentType) : string;
```
Comment (funivan) [inline src/Helper/EncodingConverterInterface.php]: invalid name for method. Better to use : `convert`

File: src/Helper/HtmlEncodingConverter.php [hunk 4]
```diff
@@ -1,30 +1,31 @@
...
+
   namespace Xparse\Parser\Helper;
 
   /**
-   * Class EncodingConverter
-   * @package Xparse\Parser\Helper
+   * Try to convert input encoding
    */
-  class HtmlEncodingConverter {
+  class HtmlEncodingConverter implements EncodingConverterInterface {
```
Comment (funivan) [inline src/Helper/HtmlEncodingConverter.php]: Invalid name for class. Better name `ToUtfConverter`

File: src/Helper/HtmlEncodingConverter.php [hunk 5]
```diff
@@ -1,30 +1,31 @@
...
     /**
-     * @var array|null
+     * @var array
      */
-    private static $supportedEncodings = null;
+    private $supportedEncodings;
+
 
+    public function __construct() {
+      $this->supportedEncodings = $this->getSupportedEncodings();
```
Comment (funivan) [inline src/Helper/HtmlEncodingConverter.php]: We do not need to do that job in the constructor. Use  cache technique for the method `getSupportedEncondings`

File: src/Helper/HtmlEncodingConverter.php [hunk 6]
```diff
@@ -1,30 +1,31 @@
...
-     * @return string
+     * @inheritdoc
      */
-    public static function convertToUtf($html, $contentType = false) {
+    public function convertToUtf(string $html, string $contentType = '') : string {
       $encoding = null;
-      if (!empty($contentType)) {
-        preg_match("!^.*charset=([A-Za-z0-9-]{4,})$!", $contentType, $contentTypeData);
+      if ($contentType !== '') {
+        preg_match('!^.*charset=([A-Za-z0-9-]{4,})$!', $contentType, $contentTypeData);
```
Comment (funivan) [inline src/Helper/HtmlEncodingConverter.php]: We can convert this if statement. For example: 
```php
if ( preg_match === 1 ){
$encoding = $contentTypeData[1];
}
```

File: src/Helper/HtmlEncodingConverter.php [hunk 7]
```diff
@@ -49,29 +50,26 @@ public static function convertToUtf($html, $contentType = false) {
...
      */
-    private static function getSupportedEncodings() {
-      if (self::$supportedEncodings !== null) {
-        return self::$supportedEncodings;
-      }
+    private function getSupportedEncodings() : array {
 
       $hasAliasesFunction = function_exists('mb_encoding_aliases');
-      self::$supportedEncodings = [];
+      $supportedEncodings = [];
```
Comment (funivan) [inline src/Helper/HtmlEncodingConverter.php]: we do not need such long names. just use `$encodings`

File: src/Helper/LinkConverter.php [hunk 8]
```diff
@@ -1,27 +1,27 @@
...
 
     /**
-     * Convert relative links, images scr and form actions to absolute
-     *
-     * @param ElementFinder $page
-     * @param string $affectedUrl
+     * @inheritdoc
      */
-    public static function convertUrlsToAbsolute(ElementFinder $page, $affectedUrl) {
+    public function relativeToAbsolute(ElementFinder $finder, string $affectedUrl = '') {
```
Comment (funivan) [inline src/Helper/LinkConverter.php]: just name it `convert` and rename class to `RelativeToAbosuleLinkConverter`

**Comment by ganchclub** [comment]
Yes, I think it's a good idea.

File: src/Parser.php [hunk 9]
```diff
@@ -52,67 +49,66 @@ public function __construct(ClientInterface $client = null, ElementFinderFactory
...
      */
-    public function get($url, array $options = []) {
-      if (empty($url) or !is_string($url)) {
-        throw new \InvalidArgumentException('Url must be not empty and string.');
+    public function get(string $url, array $options = []) : ElementFinder {
+      if ($url === '') {
+        throw new \InvalidArgumentException('Url can\'t be empty');
       }
 
       $request = new Request('GET', $url);
```
Comment (funivan) [inline src/Parser.php:73]: we do not need this variable here. Just inline it ;)

File: src/Parser.php [hunk 10]
```diff
@@ -52,67 +49,66 @@ public function __construct(ClientInterface $client = null, ElementFinderFactory
...
     }
 
 
     /**
-     * @param \Xparse\ElementFinder\ElementFinder $lastPage
+     * @param ElementFinder $lastPage
      * @return $this
      */
-    public function setLastPage($lastPage) {
+    private function setLastPage(ElementFinder $lastPage) : self {
```
Comment (funivan) [inline src/Parser.php:107]: we can even remove this method

File: src/Parser.php [hunk 11]
```diff
@@ -128,24 +124,23 @@ public function getLastResponse() {
...
      */
-    public function getClient() {
+    public function getClient() : ClientInterface {
       return $this->client;
     }
 
 
     /**
-     * @param $request
+     * @param RequestInterface $request
```
Comment (funivan) [inline src/Parser.php:131]: :+1:

File: src/ParserInterface.php [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+     * @param string $url
+     * @param array $options
+     * @return ElementFinder
+     */
+    public function post(string $url, array $options) : ElementFinder;
+
+    /**
+     * @return ElementFinder
+     */
+    public function getLastPage() : ElementFinder;
```
Comment (funivan) [inline src/ParserInterface.php:27]: Do we really need this method ?

**Review by funivan — CHANGES_REQUESTED** [review]

File: src/ElementFinderFactory.php (hunk 2 above)
Comment (ganchclub) [reply src/ElementFinderFactory.php]: it's used in `ElementFinderFactory::create` method

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,12 +1,13 @@
 #Changelog
 All Notable changes to `Parser` will be documented in this file
-## 0.1.0-alpha.5 [Unreleased]
+## 0.2.0 [Unreleased]
 
 ### Changed
 - `Parser::post` signature was changed, `$body` parameter was dropped
 - `Parser::setLastPage` method access level changed to private  
 - `HtmlEncodingConverter::convertToUtf` method becomes non-static  
 - Upgraded to phpunit 6.3
```
Comment (funivan) [inline CHANGELOG.md]: specify issue ID

File: src/Helper/ToUtfConverter.php [hunk 14]
```diff
@@ -50,26 +46,30 @@ public function convertToUtf(string $html, string $contentType = '') : string {
     /**
      * @return array
      */
-    private function getSupportedEncodings() : array {
+    private static function getSupportedEncodings() : array {
```
Comment (funivan) [inline src/Helper/ToUtfConverter.php]: Avoid static keyword.
```php
if($this->supportedEncodings === null ){
$this->supportedEncodings = //...
}
return $this->supportedEncodings;
```

**Review by funivan — CHANGES_REQUESTED** [review]

**Review by funivan — APPROVED** [review]

**Pull request merged**